cdap-client/src/main/java/co/cask/cdap/client/WorkflowClient.java

```diff
@@ -114,14 +156,33 @@ public WorkflowTokenDetail getWorkflowToken(Id.Run workflowRunId, String key)
    * {@link WorkflowToken.Scope#USER}
    * @param key the specified key. If null, it returns all keys in the specified {@link WorkflowToken.Scope}
    * @return {@link WorkflowTokenDetail} with the specified filters
+   * @deprecated since 4.0.0. Use {@link #getWorkflowToken(ProgramRunId, WorkflowToken.Scope, String)} instead.
    */
+  @Deprecated
   public WorkflowTokenDetail getWorkflowToken(Id.Run workflowRunId, @Nullable WorkflowToken.Scope scope,
                                               @Nullable String key)
     throws IOException, UnauthenticatedException, NotFoundException, UnauthorizedException {
+    return getWorkflowToken(workflowRunId.toEntityId(), scope, key);
+  }
+
+  /**
+   * Retrieve the {@link WorkflowToken} for the specified workflow run filtered by the specified
+   * {@link WorkflowToken.Scope} and key.
+   *
+   * @param workflowRunId the run id of the workflow
+   * @param scope the specified {@link WorkflowToken.Scope}. If null, it returns keys for
+   * {@link WorkflowToken.Scope#USER}
+   * @param key the specified key. If null, it returns all keys in the specified {@link WorkflowToken.Scope}
+   * @return {@link WorkflowTokenDetail} with the specified filters
+   */
+  public WorkflowTokenDetail getWorkflowToken(ProgramRunId workflowRunId, @Nullable WorkflowToken.Scope scope,
+                                              @Nullable String key)
+    throws IOException, UnauthenticatedException, NotFoundException, UnauthorizedException {
     String path = String.format("apps/%s/workflows/%s/runs/%s/token",
-                                workflowRunId.getProgram().getApplicationId(), workflowRunId.getProgram().getId(),
-                                workflowRunId.getId());
-    URL url = config.resolveNamespacedURLV3(workflowRunId.getNamespace(), appendScopeAndKeyToUrl(path, scope, key));
+                                workflowRunId.getApplication(), workflowRunId.getProgram(),
+                                workflowRunId.getRun());
+    URL url = config.resolveNamespacedURLV3(workflowRunId.getParent().getNamespaceId(),
```

Comment:
I guess the real fix would be to put that method in NamespacedEntityId instead of ProgramId.